designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,585 @@
+
+This file tells ESLint to load `eslint-plugin-react` and then configure a rule from that plugin. The `react/` is automatically preprended to the rule by ESLint for easy reference.
+
+In `eslint.config.js`, the same configuration is achieved using a `ruledefs` key:
+
+```js
+const reactPlugin = require("eslint-plugin-react");
+
+exports.config = {
+    files: ["*.js"],
+    ruledefs: {
+        react: reactPlugin.rules
+    },
+    rules: {
+        "react/jsx-uses-react": "error"
+    }
+};
+```
+
+Here, it is the `ruledefs` that assigns the name `react` to the rules from `eslint-plugin-react`. The reference to `react/` in a rule will always look up that value in the `ruledefs` key.
+
+#### Referencing Parsers and Processors
+
+In `.eslintrc`, the `parser` and `processor` keys required strings to be specified, such as:
+
+```yaml
+plugins: ["markdown"]
+parser: "babel-eslint"
+processor: "markdown/markdown"
+```
+
+In `eslint.config.js`, you would need to pass the references directly, such as:
+
+```js
+exports.config = {
+    files: ["*.js"],
+    parser: require("babel-eslint"),
+    processor: require("eslint-plugin-markdown").processors.markdown
+};
+```
+
+In both cases, users now must pass a direct object reference. This has the benefit of using the builtin Node.js module resolution system or allowing users to specify their own. There is never a question of where the modules will be resolved from.
+
+#### Applying an Environment
+
+Unlike with `.eslintrc` files, there is no `env` key in `eslint.config.js`. Users can mimic the behavior of `env` by assigning directly to the `globals` key:
+
+```js
+const globals = require("globals");
+
+exports.config = {
+    files: ["*.js"],
+    globals: {
+        MyGlobal: true,
+        ...globals.browser
+    }
+};
+```
+
+This effectively duplicates the use of `env: { browser: true }` in ESLint.
+
+**Note:** This would allow us to stop environments in ESLint. We could just tell people to use `globals` in their config and allow them to specify which version of `globals` they want to use.
+
+#### Extending Another Config
+
+Extending another config works the same as in `.eslintrc` except users pass entire config objects rather than the name of a config. That can easily be done in the following manner:
+
+```js
+exports.config = {
+    files: ["*.js"],
+    extends: require("eslint-config-standard"),
+    rules: {
+        semi: ["error", "always"]
+    }
+};
+```
+
+This config extends `eslint-config-standard` by assigning it to `extends`. You can also use an array for `extends` to extend from multiple configs:
+
+```js
+exports.config = {
+    files: ["*.js"],
+    extends: [
+        require("eslint-config-standard"),
+        require("@me/eslint-config")
+    ],
+    rules: {
+        semi: ["error", "always"]
+    }
+};
+```
+
+#### Extending From `eslint:recommended` and `eslint:all`
+
+Both `eslint:recommended` and `eslint:all` can be represented as strings in the `extends` key. For example:
```

Comment:
Yeah, I wanted to avoid that for a couple of reasons:

1. I don't want configs to accidentally modify those values, therefore affecting how other configs in their `extends` chain are interpreted.
2. It creates a dependency mess between configs and ESLint. In this scenario, a config could require a different version of ESLint than what the user running. By allowing strings values, we avoid this dependency issue altogether and ensure that configs aren't requiring a certain version of ESLint work correctly (though they could check the version if they wanted to).

Reply:
The first I was contemplating could be a feature, so thanks for explaining your rationale there. The second I hadn't considered, and you're completely right - that would be gross to handle.